### PR TITLE
Add Exact Matcher

### LINKED ISSUE #202: The exact matcher is only used in reference section texts
At the moment we have
```
sectioned_documents = transform_scraper_file(scraper_file)
exact_matcher = ExactMatcher(sectioned_documents, 
     settings.MATCH_TITLE_LENGTH_THRESHOLD)
```
in `refparse`, but actually for the exact matching _I think_ we should
1. search in **all** the pdf text
2. search regardless of whether the pdf has a references section or not.

### PR DESCRIPTION
**Pull request by lizgzil**
Fix #202 

This is the rebased version of https://github.com/wellcometrust/reach/pull/252

# Description

This PR adapts the code to look for exact matches in all the policy text, rather than just in the references sections.

- Adds an optional argument in `get_scraping_results` in file_manager.py which allows you to input which columns to include in the scraper output dictionary. This will mean we can include the text column needed to exact match on all the text. I did this rather than just updating the global variable `SCRAPING_COLUMNS` with 'text' since for the fuzzy match it is unneeded to store the large text data in memory.
- I create a new function `transform_scraper_text_file` in `refparse.py` to specifically transform the scraped full text for the exact match.
- Create a new dag (policy-test-exact-match) that does the exact matching (editing `exact_match_refs_operator.py` and `policy.py`).
- Adds 8 new exact matcher tests
- Changes to `test_pdf_objects` to allow for these tests to be run (or skipped) on a mac (this should be fixed properly at some point, see issue #273 )

## Type of change
- [x] :bug: Bug fix

## Running

A ran locally for the 70 documents in the latest MSF scrape, using:
```
python -m policytool.refparse.refparse \
    --scraper-file "s3://datalabs-dev/reach-airflow/output/policy/parsed-pdfs/msf/parsed-pdfs-msf.json.gz" \
    --references-file "s3://datalabs-data/wellcome_publications/uber_api_publications.csv" \
    --model-file "s3://datalabs-data/reference_parser_models/reference_parser_pipeline.pkl" \
    --output-url "file://./tmp/parser-output/output_folder_name"
```
this took **504 seconds** and finds 10 doc-publication matches.

Without these changes this took **8 seconds** and found 2 doc-publication matches (reassuring also found in the 10 full text search matches).

- I tried flashtext to make it quicker, but it didn't work on looking for whole sentences in large amounts of text - it works for 1 or 2 words in text.
- I tried Whoosh to make this quicker, but it took 1774 seconds and found 7538 matches - obviously there is probably a bug somewhere, but I didn't investigate any further.
- I tried Spacy's PhraseMatcher (`pip install spacy` and `python -m spacy download en_core_web_sm`) but aborted trying this out, since in airflow we dont need it, so it doesnt matter if refparse ran locally is slow

I've created an issue #257, if someone wants to pick it up this performance issue in the future.

# How Has This Been Tested?

```
>>> make docker-test
74 passed, 4 warnings in 10.56s
>>> make test
 72 passed, 2 skipped, 5 warnings in 7.34s
>>> cd reach/refparse/
>>> python -m unittest
Ran 35 tests in 0.112s

OK
```
I ran:
```
source build/virtualenv/bin/activate
eval $(./export_env.py)
docker-compose up -d
./docker_exec.sh airflow test policy-test ExactMatchRefs.msf 2019-11-8
```
it took about 3 hours to run and processed 5377194 publications and found 8132 matches. 
I noticed my results have duplicates in, fixing this might help speed things up if we delete duplicate policy docs before the exact matcher runs. I added some info about this duplication to issue #183

policy-test and policy-test-exact-match dags complete:
Screenshot 2019-11-14 at 14 24 09

### CODE REVIEW

File: reach/refparse/utils/exact_match.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 class ExactMatcher:
```
Comment (jdu) [inline reach/refparse/utils/exact_match.py]: Indentation is a bit wonky in this file for some reason.

See https://github.com/wellcometrust/reach/issues/257#issuecomment-555460819 about the performance of `clean_text`

Comment (hblanks) [reply reach/refparse/utils/exact_match.py]: Thanks for finding this, @jdu . I've pushed a commit that fixes this and other stray tabs I found in the repo.

**Comment by hblanks** [comment]
@nsorros , @lizgzil , @SamDepardieu , @jdu  - OK! I've updated so that:

1. The test DAG has both fuzzy and exact match, with appropriately limited inputs all the way down
1. The main DAG only has fuzzy match

And the DAG generation code is a little bit more modular along the way

This gets us to where at least the test DAG runs in short time, and we can move exact into the main DAG when we speed it up enough.

PTAL when you can, and we'll give it a merge.

**Comment by lizgzil** [comment]
@hblanks LGTM!

**Comment by hblanks** [comment]
One more update: this does increase the test DAG runtime by ~6 minutes:

gantt

@SamDepardieu , @jdu  - do let me know if you think a 15 minute runtime is acceptable. If not, we still have some options:

1. Reduce number of records matched by exact matcher in test DAG.
1. Increase parallelism (though this may be RAM intensive) by increasing `celery.worker_concurrency` from 2 to 4 in `airflow.cfg`.

**Comment by SamDepardieu** [comment]
I'm happy with a 15min job

**Comment by jdu** [comment]
15 minutes is fine as i've taken to running a full policy-test and then re-running individual tasks within it to test changes, etc... so it's not all that bad.

Maybe further down the road we can set up a `python -m HTTPServer` based site in the repo or something to have a more controlled/faster test target for the policy-test.

**Pull request merged**